### PR TITLE
test/e2e: make tests shard aware

### PR DESCRIPTION
**Pull request by s-urbaniak**
## Summary
- This forbids `system:masters` access to kcp-front-proxy
- Factors existing e2e tests to access the kcp shard as admin

## Related issue(s)

Obsoletes #1413, #1369

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -216,9 +216,9 @@ test-e2e-sharded: build-all
 	TEST_IMAGE=$$(KO_DOCKER_REPO=kind.local ko build --platform=linux/$(ARCH) ./test/e2e/fixtures/kcp-test-image) && test -n "$${TEST_IMAGE}"; \
 	NO_GORUN=1 ./bin/sharded-test-server --v=2 --log-dir-path="$(LOG_DIR)" $(TEST_SERVER_ARGS) 2>&1 & PID=$$!; echo "PID $$PID"; \
 	trap 'kill -TERM $$PID' TERM INT EXIT; \
-	while [ ! -f .kcp/admin.kubeconfig ]; do sleep 1; done; \
+	while [ ! -f .kcp-0/admin.kubeconfig ]; do sleep 1; done; \
 	NO_GORUN=1 $(GO_TEST) -race -count $(COUNT) -p $(E2E_PARALLELISM) -parallel $(E2E_PARALLELISM) $(WHAT) \
-		-args --use-default-kcp-server --syncer-image="$${SYNCER_IMAGE}" --kcp-test-image="$${TEST_IMAGE}" --pcluster-kubeconfig="$(PWD)/kind.kubeconfig" $(TEST_ARGS)
+		-args --shard-kubeconfig=$(PWD)/.kcp-0/admin.kubeconfig --root-shard-admin-context="shard:admin" --syncer-image="$${SYNCER_IMAGE}" --kcp-test-image="$${TEST_IMAGE}" --pcluster-kubeconfig="$(PWD)/kind.kubeconfig" $(TEST_ARGS)
```
Comment (sttts) [inline Makefile:221]: better `--root-shard-kubeconfig`, no?

File: cmd/sharded-test-server/frontproxy.go [hunk 2]
```diff
@@ -218,15 +218,15 @@ func kcpAdminKubeConfig(ctx context.Context, hostIP string) error {
...
 			Server:               baseHost,
 			CertificateAuthority: ".kcp/serving-ca.crt",
 		},
 	}
 	kubeConfig.Contexts = map[string]*clientcmdapi.Context{
-		"root":         {Cluster: "root", AuthInfo: "admin"},
-		"default":      {Cluster: "root:default", AuthInfo: "admin"},
-		"system:admin": {Cluster: "system:admin", AuthInfo: "admin"},
+		"root":    {Cluster: "root", AuthInfo: "kcp-admin"},
+		"default": {Cluster: "root:default", AuthInfo: "kcp-admin"},
```
Comment (sttts) [inline cmd/sharded-test-server/frontproxy.go:228]: you have to rebase? this context is gone in main.

Comment (sttts) [reply cmd/sharded-test-server/frontproxy.go:228]: oh, maybe it was just forgotten. Can you remove it? The default org is no more.

File: pkg/server/options/authentication.go [hunk 3]
```diff
@@ -35,19 +35,26 @@ import (
 	genericapiserver "k8s.io/apiserver/pkg/server"
 	"k8s.io/client-go/tools/clientcmd"
 	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+
+	"github.com/kcp-dev/kcp/pkg/authorization/bootstrap"
+)
+
+const (
+	shardAdminUserName = "shard-admin"
+	kcpAdminUserName   = "kcp-admin"
```
Comment (sttts) [inline pkg/server/options/authentication.go:44]: go doc please

File: pkg/server/options/authentication.go [hunk 4]
```diff
@@ -72,93 +79,112 @@ func (s *AdminAuthentication) AddFlags(fs *pflag.FlagSet) {
 
 	fs.StringVar(&s.KubeConfigPath, "kubeconfig-path", s.KubeConfigPath,
 		"Path to which the administrative kubeconfig should be written at startup. If this is relative, it is relative to --root-directory.")
-	fs.StringVar(&s.TokenHashFilePath, "authentication-admin-token-path", s.TokenHashFilePath,
+	fs.StringVar(&s.ShardAdminTokenHashFilePath, "authentication-admin-token-path", s.ShardAdminTokenHashFilePath,
```
Comment (sttts) [inline pkg/server/options/authentication.go:82]: needs update of the description.

File: pkg/server/options/authentication.go [hunk 5]
```diff
@@ -72,93 +79,112 @@ func (s *AdminAuthentication) AddFlags(fs *pflag.FlagSet) {
...
-		UID:    uid,
+	volatileKcpAdminToken = uuid.New().String()
+
+	shardAdminUser := &user.DefaultInfo{
+		Name:   shardAdminUserName,
+		UID:    uuid.New().String(),
 		Groups: []string{user.SystemPrivilegedGroup},
 	}
 
+	kcpAdminUser := &user.DefaultInfo{
```
Comment (sttts) [inline pkg/server/options/authentication.go:109]: go doc who this is

File: pkg/server/options/authentication.go [hunk 6]
```diff
@@ -72,93 +79,112 @@ func (s *AdminAuthentication) AddFlags(fs *pflag.FlagSet) {
...
 		}
 	}
 
-	var uid = uuid.New().String()
-	adminUser := &user.DefaultInfo{
-		Name:   user.APIServerUser,
-		UID:    uid,
+	volatileKcpAdminToken = uuid.New().String()
+
+	shardAdminUser := &user.DefaultInfo{
```
Comment (sttts) [inline pkg/server/options/authentication.go:103]: go doc who this is

File: Makefile (hunk 1 above)
Comment (s-urbaniak) [reply Makefile:221]: ok, but why the `root` prefix? the kubeconfig contains both root, and base. actually the `root-shard-admin-context` should be renamed just to `shard-admin-context`, as in kubeconfig it `system:admin` context points to cluster `base` and user `shard-admin`.

Comment (sttts) [reply Makefile:221]: root in the sense of root shard, not root workspace.

kcp-0 is the root shard.

File: test/e2e/framework/config.go [hunk 7]
```diff
@@ -57,6 +58,17 @@ func (c *testConfig) KCPKubeconfig() string {
 	}
 }
 
+func (c *testConfig) ShardKubeconfig() string {
```
Comment (sttts) [inline test/e2e/framework/config.go:61]: RootShardKubeConfig

File: test/e2e/framework/config.go [hunk 8]
```diff
@@ -23,11 +23,12 @@ import (
...
 type testConfig struct {
-	syncerImage         string
-	kcpTestImage        string
-	pclusterKubeconfig  string
-	kcpKubeconfig       string
-	useDefaultKCPServer bool
+	syncerImage                    string
+	kcpTestImage                   string
+	pclusterKubeconfig             string
+	kcpKubeconfig, shardKubeconfig string
```
Comment (sttts) [inline test/e2e/framework/config.go:29]: rootShardKubeconfig

File: test/e2e/framework/config.go [hunk 9]
```diff
@@ -57,6 +58,17 @@ func (c *testConfig) KCPKubeconfig() string {
...
 }
 
+func (c *testConfig) ShardKubeconfig() string {
+	if c.shardKubeconfig == "" {
+		return c.KCPKubeconfig()
+	}
+	return c.shardKubeconfig
+}
+
+func (c *testConfig) AdminContext() string {
```
Comment (sttts) [inline test/e2e/framework/config.go:68]: RootShardSystemAdminContext

File: cmd/sharded-test-server/frontproxy.go (hunk 2 above)
Comment (s-urbaniak) [reply cmd/sharded-test-server/frontproxy.go:228]: can you point me to what you mean with "gone" 🤔

File: test/e2e/framework/fixture.go [hunk 10]
```diff
@@ -396,7 +396,7 @@ func (sf SyncerFixture) Start(t *testing.T) *StartedSyncerFixture {
 	// Write the upstream logical cluster config to disk for the workspace plugin
 	upstreamRawConfig, err := sf.UpstreamServer.RawConfig()
 	require.NoError(t, err)
-	_, kubeconfigPath := WriteLogicalClusterConfig(t, upstreamRawConfig, sf.WorkspaceClusterName)
+	_, kubeconfigPath := WriteLogicalClusterConfig(t, upstreamRawConfig, sf.WorkspaceClusterName, TestConfig.AdminContext())
```
Comment (sttts) [inline test/e2e/framework/fixture.go:399]: falling over this. Why AdminContext? What is this kubeconfig used for?

Note (not introduced in this PR but preexisting): we should always qualify names. Kubeconfig says nothing. Is it upstream? Downstream? Is it for fake-downstream logical cluster?

File: test/e2e/framework/kcp.go [hunk 11]
```diff
@@ -373,7 +373,7 @@ func (c *kcpServer) KubeconfigPath() string {
 	return c.kubeconfigPath
 }
 
-// Config exposes a copy of the neutral client config for this server.
+// Config exposes a copy of the base client config for this server.
 func (c *kcpServer) defaultConfig() (*rest.Config, error) {
```
Comment (sttts) [inline test/e2e/framework/kcp.go:377]:
```suggestion
func (c *kcpServer) baseConfig() (*rest.Config, error) {
```

File: test/e2e/framework/kcp.go [hunk 12]
```diff
@@ -385,7 +385,7 @@ func (c *kcpServer) defaultConfig() (*rest.Config, error) {
 		return nil, err
 	}
 
-	config := clientcmd.NewNonInteractiveClientConfig(raw, "system:admin", nil, nil)
+	config := clientcmd.NewNonInteractiveClientConfig(raw, TestConfig.AdminContext(), nil, nil)
```
Comment (sttts) [inline test/e2e/framework/kcp.go:388]: now I am confused. The context we want here is the base context, no? So we need `TestConfig.BaseConfig()` See my other comments about RootShard...Context. Do we need both?

File: test/e2e/framework/fixture.go (hunk 10 above)
Comment (s-urbaniak) [reply test/e2e/framework/fixture.go:399]: i was under the impression all e2e tests have to use the same context, apparently i was wrong?

File: test/e2e/framework/kcp.go (hunk 12 above)
Comment (s-urbaniak) [reply test/e2e/framework/kcp.go:388]: see https://github.com/kcp-dev/kcp/pull/1448#discussion_r913587266

File: Makefile (hunk 1 above)
Comment (s-urbaniak) [reply Makefile:221]: ok, now this makes much more sense 👍

Comment (s-urbaniak) [reply Makefile:221]: fixed

File: pkg/server/options/authentication.go (hunk 3 above)
Comment (s-urbaniak) [reply pkg/server/options/authentication.go:44]: fixed

File: pkg/server/options/authentication.go (hunk 5 above)
Comment (s-urbaniak) [reply pkg/server/options/authentication.go:109]: fixed in the godoc above

File: pkg/server/options/authentication.go (hunk 6 above)
Comment (s-urbaniak) [reply pkg/server/options/authentication.go:103]: fixed in the godoc above

File: test/e2e/framework/config.go (hunk 8 above)
Comment (s-urbaniak) [reply test/e2e/framework/config.go:29]: fixed

File: test/e2e/framework/config.go [hunk 13]
```diff
@@ -65,6 +77,8 @@ func init() {
 
 func registerFlags(c *testConfig) {
 	flag.StringVar(&c.kcpKubeconfig, "kcp-kubeconfig", "", "Path to the kubeconfig for a kcp server.")
+	flag.StringVar(&c.shardKubeconfig, "shard-kubeconfig", "", "Path to the kubeconfig for a kcp shard server. If unset, kcp-kubeconfig is used.")
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply test/e2e/framework/config.go:80]: fixed

File: test/e2e/framework/config.go [hunk 14]
```diff
@@ -65,6 +77,8 @@ func init() {
...
+	flag.StringVar(&c.adminContext, "admin-context", "system:admin", "The context in the root-kubeconfig for a system:masters admin.")
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply test/e2e/framework/config.go:81]: fixed

File: test/e2e/framework/kcp.go (hunk 12 above)
Comment (s-urbaniak) [reply test/e2e/framework/kcp.go]: resolved, this PR now goes into the right direction

File: cmd/kcp-front-proxy/options/authentication.go [hunk 15]
```diff
@@ -41,7 +42,9 @@ type Authentication struct {
 
 // NewAuthentication creates a default Authentication
 func NewAuthentication() *Authentication {
-	return &Authentication{}
+	return &Authentication{
+		DropGroups: []string{user.SystemPrivilegedGroup},
```
Comment (sttts) [inline cmd/kcp-front-proxy/options/authentication.go:46]: this should go into its own commit. It's the core of the PR.

File: Makefile [hunk 16]
```diff
@@ -80,7 +80,7 @@ endif
 
 build: WHAT ?= ./cmd/...
 build: pre-build-checks ## Build the project
-	go build -ldflags="$(LDFLAGS)" -o bin $(WHAT)
+	go build $(BUILDFLAGS) -ldflags="$(LDFLAGS)" -o bin $(WHAT)
```
[the remark this answers is not in the document]
Comment (sttts) [reply Makefile:83]: btw, something broke here, probably due to the rebase PR. This commit includes all changes.

[the remark this answers is not in the document]
Comment (s-urbaniak) [reply Makefile:83]: fixed, broke into individual commits, the broken build was because of other changes.

File: cmd/kcp-front-proxy/options/authentication.go (hunk 15 above)
Comment (s-urbaniak) [reply cmd/kcp-front-proxy/options/authentication.go:46]: fixed

File: pkg/authorization/apibinding_authorizer.go [hunk 17]
```diff
@@ -131,11 +135,14 @@ func (a *apiBindingAccessAuthorizer) Authorize(ctx context.Context, attr authori
 
 	// If bound, create a rbac authorizer filtered to the cluster.
 	clusterKubeInformer := rbacwrapper.FilterInformers(logicalcluster.From(apiExport), a.versionedInformers.Rbac().V1())
+	mergedClusterRoleInformer := rbacwrapper.MergedClusterRoleInformer(clusterKubeInformer.ClusterRoles(), a.bootstrapInformer.ClusterRoles())
+	mergedRoleInformer := rbacwrapper.MergedRoleInformer(clusterKubeInformer.Roles(), a.bootstrapInformer.Roles())
+	mergedClusterRoleBindingsInformer := rbacwrapper.MergedClusterRoleBindingInformer(clusterKubeInformer.ClusterRoleBindings(), a.bootstrapInformer.ClusterRoleBindings())
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/authorization/apibinding_authorizer.go:140]: ah yes, that may be a remainder of a previous attempt, good catch 👍

[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/authorization/apibinding_authorizer.go]: patch removed

File: pkg/authorization/toplevel_org_authorizer.go [hunk 18]
```diff
@@ -43,13 +44,18 @@ import (
 // the top-level workspace exists, and Deny otherwise.
 func NewTopLevelOrganizationAccessAuthorizer(versionedInformers clientgoinformers.SharedInformerFactory, clusterWorkspaceLister tenancyv1.ClusterWorkspaceLister, delegate authorizer.Authorizer) authorizer.Authorizer {
 	rootKubeInformer := rbacwrapper.FilterInformers(tenancyv1alpha1.RootCluster, versionedInformers.Rbac().V1())
+	bootstrapInformer := rbacwrapper.FilterInformers(genericcontrolplane.LocalAdminCluster, versionedInformers.Rbac().V1())
+
+	mergedClusterRoleInformer := rbacwrapper.MergedClusterRoleInformer(rootKubeInformer.ClusterRoles(), bootstrapInformer.ClusterRoles())
+	mergedRoleInformer := rbacwrapper.MergedRoleInformer(rootKubeInformer.Roles(), bootstrapInformer.Roles())
+	mergedClusterRoleBindingsInformer := rbacwrapper.MergedClusterRoleBindingInformer(rootKubeInformer.ClusterRoleBindings(), bootstrapInformer.ClusterRoleBindings())
```
Comment (sttts) [inline pkg/authorization/toplevel_org_authorizer.go:51]: same comment as before ;-)

File: test/e2e/homeworkspaces/home_workspaces_test.go [hunk 19]
```diff
@@ -207,10 +207,11 @@ func TestUserHomeWorkspaces(t *testing.T) {
 			orgClusterName := framework.NewOrganizationFixture(t, server)
 
 			// create non-virtual clients
-			kcpConfig := server.DefaultConfig(t)
+			kcpConfig := server.BaseConfig(t)
+			rootShardCfg := server.RootShardConfig(t)
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply test/e2e/homeworkspaces/home_workspaces_test.go:211]: `kcpClusterClient` in fact is not used here as a general kcp cluster client, just to create a cluster workspace once in https://github.com/kcp-dev/kcp/blob/103546ff5869a7ec219d22071f024a800d9fb2c8/test/e2e/homeworkspaces/home_workspaces_test.go#L188.

so renaming to simply `rootKcpClusterClient` is good 👍

File: test/e2e/homeworkspaces/home_workspaces_test.go [hunk 20]
```diff
@@ -190,7 +190,7 @@ func TestUserHomeWorkspaces(t *testing.T) {
 						Name: "workspace1",
 					},
 				}, metav1.CreateOptions{})
-				require.EqualError(t, err, `clusterworkspaces.tenancy.kcp.dev "~" is forbidden: User "system:apiserver" cannot create resource "clusterworkspaces/workspace" in API group "tenancy.kcp.dev" at the cluster scope: workspace access not permitted`, "system:master should be not able to trigger the automatic creation of user-1 home")
+				require.EqualError(t, err, `clusterworkspaces.tenancy.kcp.dev "~" is forbidden: User "shard-admin" cannot create resource "clusterworkspaces/workspace" in API group "tenancy.kcp.dev" at the cluster scope: workspace access not permitted`, "system:master should be not able to trigger the automatic creation of user-1 home")
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply test/e2e/homeworkspaces/home_workspaces_test.go:193]: i was stumbling across this myself. maybe worth considering removing it @davidfestal ?

File: pkg/authorization/toplevel_org_authorizer.go (hunk 18 above)
Comment (s-urbaniak) [reply pkg/authorization/toplevel_org_authorizer.go:51]: ha, sorry, which one ? 😄

Comment (s-urbaniak) [reply pkg/authorization/toplevel_org_authorizer.go:51]: we removed the merging in api binding authorizer, in top level authorizer it has to stay, no?

File: test/e2e/homeworkspaces/home_workspaces_test.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply test/e2e/homeworkspaces/home_workspaces_test.go:211]: done

File: pkg/authorization/toplevel_org_authorizer.go (hunk 18 above)
Comment (sttts) [reply pkg/authorization/toplevel_org_authorizer.go:51]: nvmd. This is toplevel authz.

File: test/e2e/homeworkspaces/home_workspaces_test.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply test/e2e/homeworkspaces/home_workspaces_test.go:211]: actually, renamed to `rootShardKcpClusterClient`

File: pkg/authorization/bootstrap/policy.go [hunk 21]
```diff
@@ -24,13 +24,18 @@ import (
 
 const (
 	SystemKcpClusterWorkspaceAccessGroup = "system:kcp:clusterworkspace:access"
-	SystemKcpClusterWorkspaceAdminGroup  = "system:kcp:clusterworkspace:admin"
+	// An admin group per cluster workspace.
+	// Members of this group have all permissions in the referenced cluster workspace (capped by maximal permission policy).
+	SystemKcpClusterWorkspaceAdminGroup = "system:kcp:clusterworkspace:admin"
+	// A global admin group.
+	// Members of this group have all permissions across all cluster workspaces.
+	SystemKcpAdminGroup = "system:kcp:admin"
```
Comment (sttts) [inline pkg/authorization/bootstrap/policy.go:32]: @csams fyi, this is the new group I was mentioning. Little less than system:masters, i.e. it goes through authz.

File: pkg/server/options/authentication.go [hunk 22]
```diff
@@ -72,93 +84,115 @@ func (s *AdminAuthentication) AddFlags(fs *pflag.FlagSet) {
...
 			CertificateAuthorityData: caData,
 			TLSServerName:            tlsServerName,
 		},
 	}
 	kubeConfig.Contexts = map[string]*clientcmdapi.Context{
-		"root":         {Cluster: "root", AuthInfo: adminUserName},
-		"system:admin": {Cluster: "system:admin", AuthInfo: adminUserName},
+		"root":         {Cluster: "root", AuthInfo: kcpAdminUserName},
+		"base":         {Cluster: "base", AuthInfo: kcpAdminUserName},
+		"system:admin": {Cluster: "base", AuthInfo: shardAdminUserName},
```
Comment (sttts) [inline pkg/server/options/authentication.go:195]: @csams the changes in this file will impact your deployment

File: test/e2e/conformance/cross_logical_cluster_list_test.go [hunk 23]
```diff
@@ -92,7 +93,7 @@ func TestCrossLogicalClusterList(t *testing.T) {
 	}
 
 	t.Logf("Listing ClusterWorkspace CRs across logical clusters with identity")
-	rootCfg := framework.ShardConfig(t, kcpClients, "root", cfg)
+	rootCfg := framework.ShardConfig(t, kcpClients, "root", rootShardCfg)
```
Comment (sttts) [inline test/e2e/conformance/cross_logical_cluster_list_test.go:96]: this is strange. Can we drop this line?

File: test/e2e/conformance/cross_logical_cluster_list_test.go [hunk 24]
```diff
@@ -186,7 +188,7 @@ func TestCRDCrossLogicalClusterListPartialObjectMetadata(t *testing.T) {
 	bootstrapCRD(t, wsNormalCRD1b, crdClusterClient.Cluster(wsNormalCRD1b).ApiextensionsV1().CustomResourceDefinitions(), sheriffCRD1)
 
 	t.Logf("Create a root shard client that is able to do wildcard requests")
-	rootCfg := framework.ShardConfig(t, kcpClusterClient, "root", cfg)
+	rootCfg := framework.ShardConfig(t, kcpClusterClient, "root", rootShardConfig)
```
Comment (sttts) [inline test/e2e/conformance/cross_logical_cluster_list_test.go:191]: can we drop this line?

File: test/e2e/conformance/cross_logical_cluster_list_test.go [hunk 25]
```diff
@@ -270,12 +272,13 @@ func TestBuiltInCrossLogicalClusterListPartialObjectMetadata(t *testing.T) {
...
 
-	cfg := server.DefaultConfig(t)
+	cfg := server.BaseConfig(t)
+	rootShardCfg := server.RootShardConfig(t)
 
 	kcpClusterClient, err := kcpclientset.NewClusterForConfig(cfg)
 	require.NoError(t, err, "failed to construct kcp client for server")
 
-	rootCfg := framework.ShardConfig(t, kcpClusterClient, "root", cfg)
+	rootCfg := framework.ShardConfig(t, kcpClusterClient, "root", rootShardCfg)
```
Comment (sttts) [inline test/e2e/conformance/cross_logical_cluster_list_test.go:281]: can we drop this line?

File: test/e2e/homeworkspaces/home_workspaces_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (sttts) [reply test/e2e/homeworkspaces/home_workspaces_test.go:193]: have rewritten that a little, and now understand why it is there. You change looks fine.

File: test/e2e/reconciler/clusterworkspacedeletion/controller_test.go [hunk 26]
```diff
@@ -201,12 +201,12 @@ func TestWorkspaceDeletionController(t *testing.T) {
 
 				// get clients for the right shards. We have to access the shards directly to see object (Namespace and ClusterWorkspace) deletion
 				// without being stopped at the (front-proxy) gate because the parent workspace is already gone.
-				rootShardConfig := framework.ShardConfig(t, server.kcpClusterClient, "root", server.RunningServer.DefaultConfig(t))
+				rootShardConfig := framework.ShardConfig(t, server.kcpClusterClient, "root", server.RunningServer.BaseConfig(t))
```
Comment (sttts) [inline test/e2e/reconciler/clusterworkspacedeletion/controller_test.go:204]: can we drop this?

File: test/e2e/watchcache/watchcache_enabled_test.go [hunk 27]
```diff
@@ -62,17 +62,17 @@ func TestWatchCacheEnabledForCRD(t *testing.T) {
 	t.Cleanup(cancel)
 	org := framework.NewOrganizationFixture(t, server)
 	cluster := framework.NewWorkspaceFixture(t, server, org)
```
Comment (sttts) [inline test/e2e/watchcache/watchcache_enabled_test.go:64]: schedule to root shard via option

File: test/e2e/conformance/cross_logical_cluster_list_test.go (hunk 25 above)
Comment (s-urbaniak) [reply test/e2e/conformance/cross_logical_cluster_list_test.go:281]: ha, stumbled upon this as well. this is used once in a wildcard list call: https://github.com/kcp-dev/kcp/blob/8c833d4f55313383b8d8bc3d4069f06822df893d/test/e2e/conformance/cross_logical_cluster_list_test.go#L306

File: test/e2e/conformance/cross_logical_cluster_list_test.go (hunk 24 above)
Comment (s-urbaniak) [reply test/e2e/conformance/cross_logical_cluster_list_test.go:191]: same as the others, used once in a wildcard call: https://github.com/kcp-dev/kcp/blob/8c833d4f55313383b8d8bc3d4069f06822df893d/test/e2e/conformance/cross_logical_cluster_list_test.go#L196

File: test/e2e/reconciler/clusterworkspacedeletion/controller_test.go (hunk 26 above)
Comment (s-urbaniak) [reply test/e2e/reconciler/clusterworkspacedeletion/controller_test.go:204]: oh, this is an old name, i didn't touch it and in fact it uses `BaseConfig()`, not `RootShardConfig()`

File: test/e2e/watchcache/watchcache_enabled_test.go (hunk 27 above)
Comment (s-urbaniak) [reply test/e2e/watchcache/watchcache_enabled_test.go:64]: i am sorry, i don't follow, can you elaborate?

File: test/e2e/reconciler/clusterworkspacedeletion/controller_test.go (hunk 26 above)
Comment (sttts) [reply test/e2e/reconciler/clusterworkspacedeletion/controller_test.go:204]: nvmd. This uses BaseConfig, not rootCfg

File: test/e2e/watchcache/watchcache_enabled_test.go (hunk 27 above)
Comment (sttts) [reply test/e2e/watchcache/watchcache_enabled_test.go:64]: `framework.NewWorkspaceFixture(t, server, org, framework.WithShardConstraints(tenancyapi.ShardConstraints{Name: "root"})),`

File: test/e2e/conformance/cross_logical_cluster_list_test.go (hunk 25 above)
Comment (sttts) [reply test/e2e/conformance/cross_logical_cluster_list_test.go:281]: rootCfg will be the same as rootShardCfg. ShardConfig on rootShardCfg is a noop.

File: test/e2e/conformance/cross_logical_cluster_list_test.go (hunk 24 above)
Comment (sttts) [reply test/e2e/conformance/cross_logical_cluster_list_test.go:191]: rootCfg will be the same as rootShardCfg. ShardConfig on rootShardCfg is a noop.

File: test/e2e/conformance/cross_logical_cluster_list_test.go (hunk 25 above)
Comment (s-urbaniak) [reply test/e2e/conformance/cross_logical_cluster_list_test.go]: removed

File: test/e2e/conformance/cross_logical_cluster_list_test.go (hunk 24 above)
Comment (s-urbaniak) [reply test/e2e/conformance/cross_logical_cluster_list_test.go]: removed

File: test/e2e/conformance/cross_logical_cluster_list_test.go (hunk 23 above)
Comment (s-urbaniak) [reply test/e2e/conformance/cross_logical_cluster_list_test.go]: removed

File: test/e2e/watchcache/watchcache_enabled_test.go (hunk 27 above)
Comment (s-urbaniak) [reply test/e2e/watchcache/watchcache_enabled_test.go]: fixed

File: test/e2e/homeworkspaces/home_workspaces_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply test/e2e/homeworkspaces/home_workspaces_test.go:193]: ok

File: test/e2e/apibinding/apibinding_test.go [hunk 28]
```diff
@@ -59,11 +59,11 @@ func TestAPIBinding(t *testing.T) {
...
-	serviceProvider2Workspace := framework.NewWorkspaceFixture(t, server, orgClusterName)
-	consumer1Workspace := framework.NewWorkspaceFixture(t, server, orgClusterName)
-	consumer2Workspace := framework.NewWorkspaceFixture(t, server, orgClusterName)
-	consumer3Workspace := framework.NewWorkspaceFixture(t, server, orgClusterName)
+	serviceProvider1Workspace := framework.NewWorkspaceFixture(t, server, orgClusterName, framework.WithShardConstraints(tenancyv1alpha1.ShardConstraints{Name: "root"}))
+	serviceProvider2Workspace := framework.NewWorkspaceFixture(t, server, orgClusterName, framework.WithShardConstraints(tenancyv1alpha1.ShardConstraints{Name: "root"}))
+	consumer1Workspace := framework.NewWorkspaceFixture(t, server, orgClusterName, framework.WithShardConstraints(tenancyv1alpha1.ShardConstraints{Name: "root"}))
+	consumer2Workspace := framework.NewWorkspaceFixture(t, server, orgClusterName, framework.WithShardConstraints(tenancyv1alpha1.ShardConstraints{Name: "root"}))
+	consumer3Workspace := framework.NewWorkspaceFixture(t, server, orgClusterName, framework.WithShardConstraints(tenancyv1alpha1.ShardConstraints{Name: "root"}))
```
Comment (sttts) [inline test/e2e/apibinding/apibinding_test.go:66]: these are wrong. The root access is only needed to get the shard list. The actual test is not using root access.

Comment (s-urbaniak) [reply test/e2e/apibinding/apibinding_test.go]: oha, fixed
